### PR TITLE
BUGFIX: 4415 Fusion error on dev cache invalidation

### LINKED ISSUE #4415: BUG: `Neos_Fusion_ParsePartials` cache not flushed for moved files because `realpath` fails
From Neos 8.3.0 .. 8.3.9 we didnt handle this case by just throwing an error.

```
Exception in line 53 of /Users/marchenryschultz/Code/core/Neos.NeosIo/Packages/Application/Neos.Fusion/Classes/Core/Cache/ParserCacheFlusher.php: Couldn't resolve realpath for: '/Users/marchenryschultz/Code/core/Neos.NeosIo/Packages/Sites/Neos.NeosIo/Resources/Private/Fusion/Content/BrandLogo/BrandLogo.fusion'

21 Neos\Fusion\Core\Cache\ParserCacheFlusher::getCacheIdentifierForFile("/Users/marchenryschultz/Code/core/Neos.NeosIo/Pack…/Private/Fusion/Content/BrandLogo/BrandLogo.fusion")
20 Neos\Fusion\Core\Cache\ParserCacheFlusher::flushPartialCacheOnFileChanges("Fusion_Files", array|107|, "Neos\Flow\Monitor\FileMonitor::filesHaveChanged")
19 call_user_func_array(array|2|, array|3|)
18 Neos\Flow\SignalSlot\Dispatcher::dispatch("Neos\Flow\Monitor\FileMonitor", "filesHaveChanged", array|2|)
17 Neos\Flow\Monitor\FileMonitor_Original::emitFilesHaveChanged("Fusion_Files", array|107|)
16 Neos\Flow\Monitor\FileMonitor_Original::detectChanges()
15 Neos\Fusion\Package::Neos\Fusion\{closure}(Neos\Flow\Core\Booting\Step, "runtime", "Neos\Flow\Core\Booting\Sequence::afterInvokeStep")
14 Closure::__invoke(Neos\Flow\Core\Booting\Step, "runtime", "Neos\Flow\Core\Booting\Sequence::afterInvokeStep")
13 call_user_func_array(array|2|, array|3|)
12 Neos\Flow\SignalSlot\Dispatcher::dispatch("Neos\Flow\Core\Booting\Sequence", "afterInvokeStep", array|2|)
11 Neos\Flow\Core\Booting\Sequence::invokeStep(Neos\Flow\Core\Booting\Step, Neos\Flow\Core\Bootstrap)
10 Neos\Flow\Core\Booting\Sequence::invokeStep(Neos\Flow\Core\Booting\Step, Neos\Flow\Core\Bootstrap)
9 Neos\Flow\Core\Booting\Sequence::invokeStep(Neos\Flow\Core\Booting\Step, Neos\Flow\Core\Bootstrap)
8 Neos\Flow\Core\Booting\Sequence::invokeStep(Neos\Flow\Core\Booting\Step, Neos\Flow\Core\Bootstrap)
7 Neos\Flow\Core\Booting\Sequence::invokeStep(Neos\Flow\Core\Booting\Step, Neos\Flow\Core\Bootstrap)
6 Neos\Flow\Core\Booting\Sequence::invokeStep(Neos\Flow\Core\Booting\Step, Neos\Flow\Core\Bootstrap)
5 Neos\Flow\Core\Booting\Sequence::invokeStep(Neos\Flow\Core\Booting\Step, Neos\Flow\Core\Bootstrap)
4 Neos\Flow\Core\Booting\Sequence::invoke(Neos\Flow\Core\Bootstrap)
3 Neos\Flow\Http\RequestHandler::boot()
2 Neos\Flow\Http\RequestHandler::handleRequest()
1 Neos\Flow\Core\Bootstrap::run()
```

With Neos 8.3.10 https://github.com/neos/neos-development-collection/pull/4838 tried to fix this, but accidentally completely broke caching.

Instead the current hotfix is to silently ignore removed fusion files and not trigger cache invalidator for them.
That would lead to a growing cache and maybe odd behaviour in some cases, but that seems a better compromise.
Related issue of flows php caching not being flushed for delted classes: https://github.com/neos/flow-development-collection/issues/3303

<details><summary>for some reason it also happened for js files???</summary>
<p>

yes because we listen to everything: https://github.com/neos/neos-development-collection/issues/4606#issuecomment-1896187044

```
Exception in line 53 of /Users/marchenryschultz/Code/core/Neos.NeosIo/Packages/Application/Neos.Fusion/Classes/Core/Cache/ParserCacheFlusher.php: Couldn't resolve realpath for: '/Users/marchenryschultz/Code/core/Neos.NeosIo/Packages/Sites/Neos.NeosIo/Resources/Private/Fusion/Content/Tabs/Tabs.js'

21 Neos\Fusion\Core\Cache\ParserCacheFlusher::getCacheIdentifierForFile("/Users/marchenryschultz/Code/core/Neos.NeosIo/Pack…osIo/Resources/Private/Fusion/Content/Tabs/Tabs.js")
20 Neos\Fusion\Core\Cache\ParserCacheFlusher::flushPartialCacheOnFileChanges("Fusion_Files", array|2|, "Neos\Flow\Monitor\FileMonitor::filesHaveChanged")
19 call_user_func_array(array|2|, array|3|)
18 Neos\Flow\SignalSlot\Dispatcher::dispatch("Neos\Flow\Monitor\FileMonitor", "filesHaveChanged", array|2|)
17 Neos\Flow\Monitor\FileMonitor_Original::emitFilesHaveChanged("Fusion_Files", array|2|)
16 Neos\Flow\Monitor\FileMonitor_Original::detectChanges()
15 Neos\Fusion\Package::Neos\Fusion\{closure}(Neos\Flow\Core\Booting\Step, "runtime", "Neos\Flow\Core\Booting\Sequence::afterInvokeStep")
14 Closure::__invoke(Neos\Flow\Core\Booting\Step, "runtime", "Neos\Flow\Core\Booting\Sequence::afterInvokeStep")
13 call_user_func_array(array|2|, array|3|)
12 Neos\Flow\SignalSlot\Dispatcher::dispatch("Neos\Flow\Core\Booting\Sequence", "afterInvokeStep", array|2|)
11 Neos\Flow\Core\Booting\Sequence::invokeStep(Neos\Flow\Core\Booting\Step, Neos\Flow\Core\Bootstrap)
10 Neos\Flow\Core\Booting\Sequence::invokeStep(Neos\Flow\Core\Booting\Step, Neos\Flow\Core\Bootstrap)
9 Neos\Flow\Core\Booting\Sequence::invokeStep(Neos\Flow\Core\Booting\Step, Neos\Flow\Core\Bootstrap)
8 Neos\Flow\Core\Booting\Sequence::invokeStep(Neos\Flow\Core\Booting\Step, Neos\Flow\Core\Bootstrap)
7 Neos\Flow\Core\Booting\Sequence::invokeStep(Neos\Flow\Core\Booting\Step, Neos\Flow\Core\Bootstrap)
6 Neos\Flow\Core\Booting\Sequence::invokeStep(Neos\Flow\Core\Booting\Step, Neos\Flow\Core\Bootstrap)
5 Neos\Flow\Core\Booting\Sequence::invokeStep(Neos\Flow\Core\Booting\Step, Neos\Flow\Core\Bootstrap)
4 Neos\Flow\Core\Booting\Sequence::invoke(Neos\Flow\Core\Bootstrap)
3 Neos\Flow\Http\RequestHandler::boot()
2 Neos\Flow\Http\RequestHandler::handleRequest()
1 Neos\Flow\Core\Bootstrap::run()
```

</p>
</details> 

It seems that when deleting files we dont correctly handle them.

option A: flush all if it couldn't resolve?
option B: assume its already a full path and still calculate the cache identifier.

### PR DESCRIPTION
**Pull request by mhsdesign**
Resolves #4415

After deleting a fusion file like `BrandLogo.fusion` one will face the error after booting flow:

```
Couldn't resolve realpath for: '/absolutePath/Code/core/Neos.NeosIo/Packages/Sites/Neos.NeosIo/Resources/Private/Fusion/Content/BrandLogo/BrandLogo.fusion'
```

This is caused as `realpath` doesnt work on files which are deleted. The path is (as you can see already absolute) so we dont need realpath in general here.

**Upgrade instructions**

**Review instructions**

**Checklist**

- [ ] Code follows the PSR-2 coding style
- [ ] Tests have been created, run and adjusted as needed
- [ ] The PR is created against the [lowest maintained branch](https://www.neos.io/features/release-roadmap.html)
- [ ] Reviewer - PR Title is brief but complete and starts with `FEATURE|TASK|BUGFIX`
- [ ] Reviewer - The first section explains the change briefly for change-logs
- [ ] Reviewer - Breaking Changes are marked with `!!!` and have upgrade-instructions

### CODE REVIEW

**Comment by Sebobo** [comment]
I don't fully understand why realpath doesn't properly work in this case?
Is there no PHP way to do this properly for all OS?

**Comment by mhsdesign** [comment]
sorry i was a little spare on the details so far.

as written here https://github.com/neos/neos-development-collection/issues/4415 

the problem lies here:

> It seems that when deleting files we dont correctly handle them.

flow notifies us that some files changed, but if those are deletions, we cannot resolve their realpath because they are deleted.

The reason we use realpath at all is to have a stable cache id generation.

But flow always gives us realpaths already so we can skip `realpath`

File: Neos.Fusion/Classes/Core/Cache/ParserCacheIdentifierTrait.php [hunk 1]
```diff
@@ -26,19 +26,35 @@ private function getCacheIdentifierForDslCode(string $identifier, string $code):
...
+            }
         }
-
         $realFusionFilePathWithoutRoot = str_replace(FLOW_PATH_ROOT, '', $realPath);
         return 'file_' . md5($realFusionFilePathWithoutRoot);
     }
+
+    private static function isRealPath(string $path): bool
+    {
+        $isAbsoluteWindowsPath = PHP_OS_FAMILY === 'Windows' && preg_match('`^[a-zA-Z]:/[^/]`', $path) === 1;
```
Comment (bwaidelich) [inline Neos.Fusion/Classes/Core/Cache/ParserCacheIdentifierTrait.php:51]: IIRC we use other ways to determine the OS (`DIRECTORY_SEPARATOR` ?) and we should not mix them

**Review by bwaidelich — COMMENTED** [review]
Just some comment re OS detection.
Apart from that I didn't understand why exactly this is needed and IMO we need tests for lowlevel changes like this one

**Review by mficzel — COMMENTED** [review]
This should go into the fileUtility in Neos 9 and have test there. Since this fixes an annoying bug i think we can merge this here without extra tests.

File: Neos.Fusion/Classes/Core/Cache/ParserCacheIdentifierTrait.php (hunk 1 above)
Comment (mhsdesign) [reply Neos.Fusion/Classes/Core/Cache/ParserCacheIdentifierTrait.php:51]: We discussed that this is fine and we will gradually move towards this more readable check.

File: Neos.Fusion/Classes/Core/Cache/ParserCacheIdentifierTrait.php [hunk 2]
```diff
@@ -26,19 +26,35 @@ private function getCacheIdentifierForDslCode(string $identifier, string $code):
...
+        $isAbsolutePath = $path[0] === '/' || $isAbsoluteWindowsPath;
+        if (!$isAbsolutePath) {
+            return false;
+        }
+        $hasWindowsDirectoryTraversal = PHP_OS_FAMILY === 'Windows' && (str_contains($path, '\\..\\') || str_contains($path, '\\.\\'));
+        $hasDirectoryTraversal = str_contains($path, '/../') || str_contains($path, '/./') || $hasWindowsDirectoryTraversal;
+        return !$hasDirectoryTraversal;
+    }
```
Comment (mficzel) [inline Neos.Fusion/Classes/Core/Cache/ParserCacheIdentifierTrait.php:59]: I suggest to use 
```php
if (PHP_OS_FAMILY === 'Windows') {
```

File: Neos.Fusion/Classes/Core/Cache/ParserCacheIdentifierTrait.php [hunk 3]
```diff
@@ -26,19 +26,35 @@ private function getCacheIdentifierForDslCode(string $identifier, string $code):
...
+    private static function isRealPath(string $path): bool
+    {
+        $isAbsoluteWindowsPath = PHP_OS_FAMILY === 'Windows' && preg_match('`^[a-zA-Z]:/[^/]`', $path) === 1;
+        $isAbsolutePath = $path[0] === '/' || $isAbsoluteWindowsPath;
+        if (!$isAbsolutePath) {
+            return false;
+        }
+        $hasWindowsDirectoryTraversal = PHP_OS_FAMILY === 'Windows' && (str_contains($path, '\\..\\') || str_contains($path, '\\.\\'));
+        $hasDirectoryTraversal = str_contains($path, '/../') || str_contains($path, '/./') || $hasWindowsDirectoryTraversal;
+        return !$hasDirectoryTraversal;
```
Comment (mhsdesign) [inline Neos.Fusion/Classes/Core/Cache/ParserCacheIdentifierTrait.php:58]:
```suggestion
        if (PHP_OS_FAMILY === 'Windows') {
            $hasDirectoryTraversal = str_contains($path, '\\..\\') || str_contains($path, '\\.\\');
            return !$hasDirectoryTraversal;
        }
        $hasDirectoryTraversal = str_contains($path, '/../') || str_contains($path, '/./');
        return !$hasDirectoryTraversal;
```

**Comment by kitsunet** [comment]
This seems like something that exploded into things it shouldn't ? why does a non existing path even get into this method, if I want a cache identifier for a path I should know beforehand if it exists, secondly why do we need to check all of this and don't just try realpath and check the result ?

I don't think these checks should exist in the first place and probably not in the Files utility either.

**Comment by mhsdesign** [comment]
Superseded by alternative better attempt in https://github.com/neos/neos-development-collection/pull/4838

Thanks @kitsunet for the discussion ^^

**Pull request closed without merging**